[pnpm change] hard close

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

11.21.0

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

1. `pnpm change`
2. Cancel with ctrl + c

### Describe the Bug

When canceling the `pnpm change` interface, it throws a stacktrace, which is a bit ugly.

```
[ERROR] User force closed the prompt with SIGINT

pnpm: User force closed the prompt with SIGINT
    at Interface.sigint (file:///Users/my-user/.cache/node/corepack/v1/pnpm/11.21.0/dist/pnpm.mjs:160128:34)
    at Interface.emit (node:events:508:28)
    at [_ttyWrite] [as _ttyWrite] (node:internal/readline/interface:1327:18)
    at ReadStream.onkeypress (node:internal/readline/interface:284:20)
    at ReadStream.emit (node:events:520:35)
    at emitKeys (node:internal/readline/utils:371:14)
    at emitKeys.next (<anonymous>)
    at ReadStream.onData (node:internal/readline/emitKeypressEvents:64:36)
    at ReadStream.emit (node:events:508:28)
    at addChunk (node:internal/streams/readable:563:12)
```

### Expected Behavior

Softer close, nicer message, no stack trace

### Which Node.js version are you using?

24.14.1

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/change-prompt-cancel.md
+++ b/.changeset/change-prompt-cancel.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/releasing.commands": patch
+"pnpm": patch
+---
+
+Canceling a `pnpm change` prompt with Ctrl-c no longer prints a stack trace. It reports `Change canceled` and exits with a success status, like the other interactive commands [#13814](https://github.com/pnpm/pnpm/issues/13814).

--- a/pnpm11/releasing/commands/src/change/index.ts
+++ b/pnpm11/releasing/commands/src/change/index.ts
@@ -1,6 +1,7 @@
 import { checkbox, input, Separator } from '@inquirer/prompts'
 import type { Config } from '@pnpm/config.reader'
 import { PnpmError } from '@pnpm/error'
+import { globalInfo } from '@pnpm/logger'
 import {
   assembleReleasePlan,
   BUMP_TYPES,
@@ -84,7 +85,17 @@ export async function handler (opts: ChangeCommandOptions, params: string[]): Pr
   if (params.length === 1 && params[0] === 'status' && opts.bump == null && opts.summary == null) {
     return renderStatus(workspaceDir, opts)
   }
-  return recordChange(workspaceDir, opts, params)
+  try {
+    return await recordChange(workspaceDir, opts, params)
+  } catch (err: unknown) {
+    // Cancelling a prompt with Ctrl-c is how the user abandons the intent,
+    // not an error: report it and leave with a success status.
+    if (err instanceof Error && err.name === 'ExitPromptError') {
+      globalInfo('Change canceled')
+      process.exit(0)
+    }
+    throw err
+  }
 }
 
 async function recordChange (workspaceDir: string, opts: ChangeCommandOptions, params: string[]): Promise<string> {

--- a/pnpm11/releasing/commands/src/change/index.ts
+++ b/pnpm11/releasing/commands/src/change/index.ts
@@ -1,3 +1,5 @@
+import util from 'node:util'
+
 import { checkbox, input, Separator } from '@inquirer/prompts'
 import type { Config } from '@pnpm/config.reader'
 import { PnpmError } from '@pnpm/error'
@@ -88,9 +90,7 @@ export async function handler (opts: ChangeCommandOptions, params: string[]): Pr
   try {
     return await recordChange(workspaceDir, opts, params)
   } catch (err: unknown) {
-    // Cancelling a prompt with Ctrl-c is how the user abandons the intent,
-    // not an error: report it and leave with a success status.
-    if (err instanceof Error && err.name === 'ExitPromptError') {
+    if (util.types.isNativeError(err) && err.name === 'ExitPromptError') {
       globalInfo('Change canceled')
       process.exit(0)
     }

--- a/pnpm11/releasing/commands/test/change/cancel.test.ts
+++ b/pnpm11/releasing/commands/test/change/cancel.test.ts
@@ -1,0 +1,65 @@
+import fs from 'node:fs'
+import os from 'node:os'
+import path from 'node:path'
+
+import { afterEach, beforeEach, expect, it, jest } from '@jest/globals'
+
+jest.unstable_mockModule('@inquirer/prompts', () => {
+  class Separator {
+    separator: string
+    readonly type = 'separator' as const
+    constructor (separator: string) {
+      this.separator = separator
+    }
+  }
+  return {
+    Separator,
+    checkbox: jest.fn(),
+    confirm: jest.fn(),
+    input: jest.fn(),
+    password: jest.fn(),
+    select: jest.fn(),
+  }
+})
+const { checkbox } = await import('@inquirer/prompts')
+const { change } = await import('../../src/index.js')
+
+const mockCheckbox = jest.mocked(checkbox)
+
+let tempDir: string
+
+beforeEach(() => {
+  tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'pnpm-change-cancel-test-'))
+  fs.writeFileSync(path.join(tempDir, 'pnpm-workspace.yaml'), 'packages:\n  - packages/*\n')
+})
+
+afterEach(() => {
+  fs.rmSync(tempDir, { recursive: true, force: true })
+  mockCheckbox.mockReset()
+})
+
+it('change leaves without an error when the prompt is canceled', async () => {
+  const rootDir = path.join(tempDir, 'packages', 'lib')
+  fs.mkdirSync(rootDir, { recursive: true })
+  const manifest = { name: 'lib', version: '1.0.0' }
+  fs.writeFileSync(path.join(rootDir, 'package.json'), JSON.stringify(manifest, null, 2))
+
+  const canceled = new Error('User force closed the prompt')
+  canceled.name = 'ExitPromptError'
+  mockCheckbox.mockRejectedValue(canceled)
+  // `process.exit()` never returns in production, so the spy throws to stop
+  // the handler where the real call would.
+  const exited = new Error('process.exit')
+  const exitSpy = jest.spyOn(process, 'exit').mockImplementation(() => {
+    throw exited
+  })
+
+  try {
+    await expect(
+      change.handler({ dir: tempDir, workspaceDir: tempDir, allProjects: [{ rootDir, manifest }] } as any, []) // eslint-disable-line @typescript-eslint/no-explicit-any
+    ).rejects.toBe(exited)
+    expect(exitSpy).toHaveBeenCalledWith(0)
+  } finally {
+    exitSpy.mockRestore()
+  }
+})

--- a/pnpm11/releasing/commands/test/change/cancel.test.ts
+++ b/pnpm11/releasing/commands/test/change/cancel.test.ts
@@ -21,45 +21,70 @@ jest.unstable_mockModule('@inquirer/prompts', () => {
     select: jest.fn(),
   }
 })
-const { checkbox } = await import('@inquirer/prompts')
+const { checkbox, input } = await import('@inquirer/prompts')
 const { change } = await import('../../src/index.js')
 
 const mockCheckbox = jest.mocked(checkbox)
+const mockInput = jest.mocked(input)
 
 let tempDir: string
+let opts: object
 
 beforeEach(() => {
   tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'pnpm-change-cancel-test-'))
   fs.writeFileSync(path.join(tempDir, 'pnpm-workspace.yaml'), 'packages:\n  - packages/*\n')
+  const rootDir = path.join(tempDir, 'packages', 'lib')
+  fs.mkdirSync(rootDir, { recursive: true })
+  const manifest = { name: 'lib', version: '1.0.0' }
+  fs.writeFileSync(path.join(rootDir, 'package.json'), JSON.stringify(manifest, null, 2))
+  opts = { dir: tempDir, workspaceDir: tempDir, allProjects: [{ rootDir, manifest }] }
 })
 
 afterEach(() => {
   fs.rmSync(tempDir, { recursive: true, force: true })
   mockCheckbox.mockReset()
+  mockInput.mockReset()
 })
 
-it('change leaves without an error when the prompt is canceled', async () => {
-  const rootDir = path.join(tempDir, 'packages', 'lib')
-  fs.mkdirSync(rootDir, { recursive: true })
-  const manifest = { name: 'lib', version: '1.0.0' }
-  fs.writeFileSync(path.join(rootDir, 'package.json'), JSON.stringify(manifest, null, 2))
+function exitPromptError (): Error {
+  const err = new Error('User force closed the prompt')
+  err.name = 'ExitPromptError'
+  return err
+}
 
-  const canceled = new Error('User force closed the prompt')
-  canceled.name = 'ExitPromptError'
-  mockCheckbox.mockRejectedValue(canceled)
-  // `process.exit()` never returns in production, so the spy throws to stop
-  // the handler where the real call would.
+// `process.exit()` never returns in production, so the spy throws to stop the
+// handler where the real call would.
+async function expectExit (): Promise<void> {
   const exited = new Error('process.exit')
   const exitSpy = jest.spyOn(process, 'exit').mockImplementation(() => {
     throw exited
   })
-
   try {
-    await expect(
-      change.handler({ dir: tempDir, workspaceDir: tempDir, allProjects: [{ rootDir, manifest }] } as any, []) // eslint-disable-line @typescript-eslint/no-explicit-any
-    ).rejects.toBe(exited)
+    await expect(change.handler(opts as any, [])).rejects.toBe(exited) // eslint-disable-line @typescript-eslint/no-explicit-any
     expect(exitSpy).toHaveBeenCalledWith(0)
   } finally {
     exitSpy.mockRestore()
   }
+}
+
+it('change leaves without an error when the package prompt is canceled', async () => {
+  mockCheckbox.mockRejectedValue(exitPromptError())
+  await expectExit()
+})
+
+it('change leaves without an error when a bump prompt is canceled', async () => {
+  mockCheckbox.mockResolvedValueOnce(['lib']).mockRejectedValue(exitPromptError())
+  await expectExit()
+})
+
+it('change leaves without an error when the summary prompt is canceled', async () => {
+  mockCheckbox.mockResolvedValueOnce(['lib']).mockResolvedValue([])
+  mockInput.mockRejectedValue(exitPromptError())
+  await expectExit()
+})
+
+it('change rethrows a prompt failure that is not a cancellation', async () => {
+  const failed = new Error('the terminal is not interactive')
+  mockCheckbox.mockRejectedValue(failed)
+  await expect(change.handler(opts as any, [])).rejects.toBe(failed) // eslint-disable-line @typescript-eslint/no-explicit-any
 })

